bpf: Reply NA when recv ND for local IPv6 endpoints

anfernee opened the pull request:
When receiving ICMPv6 ND message from native NIC, lookup local endpoints
for target IPv6 address. If local endpoint was found, reply NA message
via the same NIC.

This enables pod's IPv6 address discoverable on the subnet to which the
node is connected to.

Torwards #10935
Signed-off-by: Yongkun Gui <ygui@google.com>

Please ensure your pull request adheres to the following guidelines:

- [ x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->

Fixes: #10935

```release-note
Add ND responder on native NIC to make endpoint IPv6 address discoverable on node network.
```

TODO: test still to be added.

pchaigno commented on bpf/init.sh:
```diff
@@ -672,6 +681,18 @@ for iface in $(ip -o -a l | awk '{print $2}' | cut -d: -f1 | cut -d@ -f1 | grep
+	*)
+		COPTS=""
+		CALLS_MAP="cilium_calls_netdev_ns_${ID_HOST}"
+		bpf_load $NATIVE_DEVS "$COPTS" "ingress" bpf_netdev.c bpf_netdev.o from-netdev $CALLS_MAP
```
I don't think adding a new BPF program is the right way to fix this. We already attach BPF programs to native devices from Golang (cf. [`pkg/datapath/loader/loader.go`](https://github.com/cilium/cilium/blob/c216dfb3ad911ca5ba7659b6266109c8c6882d70/pkg/datapath/loader/loader.go#L195)), so this will overwrite them. Doing so will also bypass the host firewall and break NodePort.

Instead, we should probably extend `icmp6_handle()`, which is already called from `bpf_host`:
https://github.com/cilium/cilium/blob/c216dfb3ad911ca5ba7659b6266109c8c6882d70/bpf/bpf_host.c#L370

Please see #12049 though. The `icmp6_handle()` call in `bpf_host` is currently broken (not compiled in) and the first commit in that pull request fixes it. The bug was introduced in v1.8 and we will backport that PR there.

anfernee replied:
I noticed `bpf_host` is broken, too. glad that you are fixing it. I'll look into #12049 too. 

I also updated the `icmp6_handle()`. My intention is for the node to respond to the NS msg from external so I attached it on `native_dev`.  If I have questions, I'll ping you on slack.

pchaigno replied:
If you try #12049 out, it might address your concern. In its current state, it will fix `bpf_host` to let all ICMPv6 NS/NA messages up the stack (in accordance with RFC4890).

anfernee replied:
I've tried #12049, the pods won't receive NS message, because the [destination IP](https://en.wikipedia.org/wiki/Solicited-node_multicast_address) is not the pod IP. I still need this but I will rebase it on top of #12049 . I will remove the new ebpf program and reuse `bpf_host`.

pchaigno reviewed (changes requested):
Why is passing ICMPv6 ND packets to the stack not enough? Is it because these packets may be dropped by pod ingress policies?

aanm commented on contrib/vagrant/start.sh:
```diff
@@ -335,6 +335,7 @@ function write_cilium_cfg() {
     fi
 
     cilium_options+=" ${TUNNEL_MODE_STRING}"
+    cilium_options+=" --device enp0s8"
```
Why this change? the `device` option has been deprecated.

anfernee replied:
Ah, It should be `--devices`..

anfernee committed:
bpf: Reply NA when recv ND for local IPv6 endpoints

When receiving ICMPv6 ND message from native NIC, lookup local endpoints
for target IPv6 address. If local endpoint was found, reply NA message
via the same NIC.

This enables pod's IPv6 address discoverable on the subnet to which the
node is connected to.

Torwards #10935
Signed-off-by: Yongkun Gui <ygui@google.com>

qmonnet commented on bpf/lib/icmp6.h:
```diff
@@ -154,8 +155,17 @@ static __always_inline int icmp6_send_echo_reply(struct __ctx_buff *ctx,
+ * @ctx:	   socket buffer
+ * @nh_off:	   offset to the IPv6 header
+ * @mac:       device mac address
+ * @to_router: ndisc is sent to router, otherwise ndisc is sent to an endpoint.
```
Nit: alignment differ between the first two and the last two arguments.

qmonnet commented on bpf/lib/icmp6.h:
```diff
@@ -347,18 +364,25 @@ static __always_inline int icmp6_send_time_exceeded(struct __ctx_buff *ctx,
+	} else if (ep) {
+		union macaddr node_mac = NODE_MAC;
+
+		return send_icmp6_ndisc_adv(ctx, nh_off, &node_mac, false);
```
```suggestion
		union macaddr node_MAC = NODE_MAC;
		
		return send_icmp6_ndisc_adv(ctx, nh_off, &node_mac, !ep);
```

qmonnet replied:
Woops, I'm sorry, I misread the conditions on this block and made a bad suggestion here :(.

What I had in mind was to condense the whole `if`/`else if` block. Reading again, it could be written like this:
```c
bool router_target = ipv6_addrcmp(&target, &router);

if (router_target || ep) {
	union macaddr node_mac = NODE_MAC;

	return send_icmp6_ndisc_adv(ctx, nh_off, &node_mac, router_target);
}
```

Otherwise, your initial version was fine, too. But my suggestion of passing `!ep` to `send_icmp6_ndisc_adv()` on the two branches was wrong (you did it just for the second one which at least is technically correct, but not what we should do). Apologies about that. Would you mind changing again, please? (Either this suggestion or your original version, as you prefer.)

qmonnet commented on bpf/lib/overloadable_skb.h:
```diff
@@ -73,9 +73,16 @@ redirect_self(const struct __sk_buff *ctx)
 #ifdef ENABLE_HOST_REDIRECT
 	return redirect(ctx->ifindex, 0);
+#else
+#ifdef NATIVE_IFINDEX
```
`#elif NATIVE_IFINDEX`?

qmonnet commented on bpf/lib/eps.h:
```diff
@@ -10,16 +10,22 @@
 
 static __always_inline __maybe_unused struct endpoint_info *
-lookup_ip6_endpoint(struct ipv6hdr *ip6)
+__lookup_ip6_endpoint(const union v6addr *ip6)
```
Could we find another function name? The leading `__` tends to make me believe this function shouldn't be called from outside of `lookup_ip6_endpoint()`.

Maybe something along `lookup_ip6_endpoint_from_daddr()`?

qmonnet reviewed (changes requested):
Couple of nits/suggestions from my side. @pchaigno, was your concern addressed?

pchaigno commented on bpf/lib/icmp6.h:
```diff
@@ -347,18 +364,25 @@ static __always_inline int icmp6_send_time_exceeded(struct __ctx_buff *ctx,
 
 	cilium_dbg(ctx, DBG_ICMP6_NS, target.p3, target.p4);
 
+	ep = lookup_ip6_endpoint_from_daddr(&target);
```
Should we only perform this lookup in the else case (i.e., `ipv6_addrcmp(&target, &router) != 0`)?

anfernee replied:
The code is changed to single branch, how about this:
```
        ep = lookup_ip6_endpoint_from_daddr(&target);

        BPF_V6(router, ROUTER_IP);
        router_target = ipv6_addrcmp(&target, &router);

        if (router_target || ep) {
                union macaddr router_mac = NODE_MAC;

                return send_icmp6_ndisc_adv(ctx, nh_off, &router_mac, router_target);
        }

```

pchaigno replied:
My main concern is that we're doing a map lookup that's unnecessary when `router_target` is true.

pchaigno commented on bpf/lib/icmp6.h:
```diff
@@ -171,9 +181,16 @@ static __always_inline int send_icmp6_ndisc_adv(struct __ctx_buff *ctx,
+		icmp6hdr.icmp6_override = 0;
+	} else {
+		icmp6hdr.icmp6_router = 0;
+		icmp6hdr.icmp6_solicited = 0;
```
Shouldn't this flag be 1 since we're answering an NS message?

pchaigno commented on bpf/lib/icmp6.h:
```diff
@@ -171,9 +181,16 @@ static __always_inline int send_icmp6_ndisc_adv(struct __ctx_buff *ctx,
+	if (to_router) {
+		icmp6hdr.icmp6_router = 1;
+		icmp6hdr.icmp6_solicited = 1;
+		icmp6hdr.icmp6_override = 0;
```
Not related to this PR, but I think this bit should be set. From the [RFC](https://tools.ietf.org/html/rfc4861#page-23):
```
                                It SHOULD NOT be
set in solicited advertisements for anycast
addresses and in solicited proxy advertisements.
It SHOULD be set in other solicited advertisements
and in unsolicited advertisements.
```

pchaigno commented on bpf/lib/icmp6.h:
```diff
@@ -347,18 +364,25 @@ static __always_inline int icmp6_send_time_exceeded(struct __ctx_buff *ctx,
+	} else if (ep) {
+		union macaddr node_mac = NODE_MAC;
+
+		return send_icmp6_ndisc_adv(ctx, nh_off, &node_mac, !ep);
```
```suggestion
		return send_icmp6_ndisc_adv(ctx, nh_off, &node_mac, false);
```

pchaigno commented on bpf/lib/overloadable_skb.h:
```diff
@@ -73,6 +73,11 @@ redirect_self(const struct __sk_buff *ctx)
+	if (ctx->ifindex == NATIVE_IFINDEX)
+		return redirect(ctx->ifindex, 0);
+	else
+		return redirect(ctx->ifindex, BPF_F_INGRESS);
```
Why is this required?

anfernee replied:
trying to keep the old behavior as much as possible. I am worried to break old logic.

pchaigno replied:
I mean why is any change to this file required? You are not even calling `redirect_self()`, right?

anfernee replied:
if does call `redirect_self()`: `send_icmp6_ndisc_adv => icmp6_send_reply => redirect_self`

pchaigno replied:
I still don't understand why this change is needed.
1. I couldn't find any definition of `NATIVE_IFINDEX` in the source code. Where is it defined? Should it be `NATIVE_DEV_IFINDEX`?
2. Assuming `NATIVE_IFINDEX` exists and is the index of a native device, why do we need to redirect packets to the _ingress_ of that device?

anfernee replied:
Originally it was only doing `redirect(ctx->ifindex, BPF_F_INGRESS);`, actually I don't understand which code path will hit this. I am guessing it could be `cilium_host` and redirect the packet back to the pods. Now the patch basically added a direction. If the packet is coming from native interface, redirect as egress, otherwise still ingress.

anfernee replied:
This is not useful after rebasing on master. I've removed it.

pchaigno commented on bpf/lib/icmp6.h:
```diff
@@ -428,6 +452,10 @@ icmp6_host_handle(struct __ctx_buff *ctx __maybe_unused)
 
+	type = icmp6_load_type(ctx, ETH_HLEN);
+	if (type == ICMP6_NS_MSG_TYPE)
+		return icmp6_handle_ns(ctx, ETH_HLEN, METRIC_INGRESS);
```
In this case, we will bypass the host firewall, but that's okay since that's what we do for NS messages anyway. However, could you switch the comment `CTX_ACT_OK` in the table below to `icmp6_handle_ns` for `ICMPv6-neighbor-solicit`?

pchaigno reviewed (dismissed):
> The reason is that the ND packet's destination address is a multicast address, which cilium doesn't recognize. The ebpf program could look into the packet and check the target address.

I'm not very familiar with these ICMPv6 mechanisms, so I'm probably just missing something, but I don't understand why the Linux stack can't handle that for us? Couldn't we teach it about pod IPv6 addresses so that it's able to answer NS messages to the node?

qmonnet reviewed (approved):
Good from my side, thanks.

joestringer commented on bpf/lib/icmp6.h:
```diff
@@ -355,10 +374,19 @@ static __always_inline int __icmp6_handle_ns(struct __ctx_buff *ctx, int nh_off)
-	if (ipv6_addrcmp(&target, &router) == 0) {
+	router_target = ipv6_addrcmp(&target, &router);
+
+	if (router_target) {
```
Isn't this condition the opposite of what the previous code tested for?

anfernee replied:
Ah.. sorry, it is somehow mistakenly changed.

joestringer reviewed (changes requested):
One conditional below seems to be swapped to the opposite value.

joestringer commented on bpf/lib/icmp6.h:
```diff
@@ -355,10 +374,19 @@ static __always_inline int __icmp6_handle_ns(struct __ctx_buff *ctx, int nh_off)
-	if (ipv6_addrcmp(&target, &router) == 0) {
+	router_target = ipv6_addrcmp(&target, &router);
+
+	if (!router_target) {
```
nit: Is there any reason to change these lines from the previous version?

`ipv6_addrcmp()` returns a u64, which is a bit unusual to convert into a bool. OTOH, the existing style of performing a comparison and checking against `0` is fairly standard compare logic in C.

joestringer replied:
I might add here, this condition is actually the case where the target is the same as the router IP so checking `!router_target` implies that this is the case that *doesn't* handle the router target, but given the final bool passed to `send_icmp6_ndisc_adv()` below, we can infer that this really is the path that handles `to_router`. So probably worth clearing up this potential confusion for future readers.

anfernee replied:
I think I copied from 
https://github.com/cilium/cilium/blob/8a03d54da11a1c15302f4f90de8bf1e539252784/bpf/lib/icmp6.h#L417

but I am you are right adding a `bool` here is not necessary. I will change it back to original shape.

joestringer reviewed (changes requested):
One more minor nit, not sure the full impact of the conversion of `u64` to `bool` but IMO the previous version of the below statement was easier to read with the prior code.

With that changed we're good to merge.
